Reduce use of downcast<>() in SVG code

cdumez opened the pull request:
#### b17ef60d82a20024e5e3ad6a1e71e1100222f6ab
<pre>
Reduce use of downcast&lt;&gt;() in SVG code
<a href="https://bugs.webkit.org/show_bug.cgi?id=266678">https://bugs.webkit.org/show_bug.cgi?id=266678</a>

Reviewed by Simon Fraser.

Reduce use of downcast&lt;&gt;() in SVG code. Use dynamicDowncast&lt;&gt;() instead.

* Source/WebCore/svg/DocumentSVG.cpp:
(WebCore::DocumentSVG::rootElement):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::createElementRenderer):
(WebCore::SVGAElement::defaultEventHandler):
* Source/WebCore/svg/SVGAltGlyphDefElement.cpp:
(WebCore::SVGAltGlyphDefElement::hasValidGlyphElements const):
* Source/WebCore/svg/SVGAltGlyphElement.cpp:
(WebCore::SVGAltGlyphElement::hasValidGlyphElements const):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::shouldApplyPathClipping const):
* Source/WebCore/svg/SVGComponentTransferFunctionElement.cpp:
(WebCore::SVGComponentTransferFunctionElement::svgAttributeChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::correspondingUseElement const):
(WebCore::SVGElement::childShouldCreateRenderer const):
(WebCore::SVGElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::buildPendingResource):
* Source/WebCore/svg/SVGFELightElement.cpp:
(WebCore::SVGFELightElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::childShouldCreateRenderer const):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(isType):
* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::rebuildFontFace):
* Source/WebCore/svg/SVGFontFaceFormatElement.cpp:
(WebCore::SVGFontFaceFormatElement::childrenChanged):
* Source/WebCore/svg/SVGFontFaceSrcElement.cpp:
(WebCore::SVGFontFaceSrcElement::childrenChanged):
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:
(WebCore::SVGFontFaceUriElement::childrenChanged):
* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
(WebCore::SVGGeometryElement::isPointInFill):
(WebCore::SVGGeometryElement::isPointInStroke):
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::toViewportElement):
(WebCore::SVGLocatable::nearestViewportElement):
(WebCore::SVGLocatable::farthestViewportElement):
(WebCore::isViewportElement): Deleted.
* Source/WebCore/svg/SVGTextPathElement.cpp:
(WebCore::SVGTextPathElement::buildPendingResource):

Canonical link: <a href="https://commits.webkit.org/272374@main">https://commits.webkit.org/272374@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/257b5c024e1ae92237e196df294687c834b7ae1a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33947 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/28497 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7377 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28118 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31794 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28079 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7340 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27986 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35290 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28588 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28436 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33634 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7576 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5601 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31477 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9237 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8267 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4108 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8086 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/svg/SVGFontFaceElement.cpp:
```diff
@@ -269,11 +269,11 @@ void SVGFontFaceElement::rebuildFontFace()
-        m_fontElement = downcast<SVGFontElement>(parentNode());
+    RefPtr<CSSValueList> list;
+    if (parentSVGFontElement) {
+        m_fontElement = parentSVGFontElement;
```
I think this can even simplified more like this:
```
m_fontElement = dynamicDowncast<SVGFontElement>(*parentNode());
if (m_fontElement)
    list = CSSValueList::createCommaSeparated(CSSFontFaceSrcLocalValue::create(AtomString { fontFamily() }));
else if (srcElement)
    list = srcElement->createSrcValue();
```

cdumez replied:
Nice indeed. Will update.

shallawa commented on Source/WebCore/svg/SVGAltGlyphDefElement.cpp:
```diff
@@ -106,13 +106,15 @@ bool SVGAltGlyphDefElement::hasValidGlyphElements(Vector<String>& glyphNames) co
-                return true;
+                // As the spec says "The first 'altGlyphItem' in which all referenced glyphs
+                // are available is chosen."
+                if (AltGlyphItemElement->hasValidGlyphElements(glyphNames) && !glyphNames.isEmpty())
```
Should not it be named `altGlyphItemElement`? The comment above is using 'altGlyphItem'.

shallawa commented on Source/WebCore/svg/SVGAElement.cpp:
```diff
@@ -128,9 +128,8 @@ void SVGAElement::defaultEventHandler(Event& event)
-                RefPtr targetElement = treeScope().getElementById(url.substringSharingImpl(1));
-                if (is<SVGSMILElement>(targetElement)) {
-                    downcast<SVGSMILElement>(*targetElement).beginByLinkActivation();
+                if (RefPtr targetElement = dynamicDowncast<SVGSMILElement>(treeScope().getElementById(url.substringSharingImpl(1)))) {
```
This became hard to read. I think separating getting the `targetElement` from the casting was clearer.

cdumez replied:
Ok, but then I either have to use a raw pointer for the downcasted pointer (which isn't great given our new smart pointer idioms) or use a still use a RefPtr for the downcasted pointer but then we do unnecessary refcounting churn that might hurt perf. 

I was trying to write safe and efficient code, which is why I changed it. How strongly do you feel about this now that you know my reasoning?

shallawa commented on Source/WebCore/svg/SVGAElement.cpp:
```diff
@@ -105,7 +105,7 @@ void SVGAElement::svgAttributeChanged(const QualifiedName& attrName)
 RenderPtr<RenderElement> SVGAElement::createElementRenderer(RenderStyle&& style, const RenderTreePosition&)
 {
-    if (is<SVGElement>(parentNode()) && downcast<SVGElement>(*parentNode()).isTextContent())
+    if (auto* svgParent = dynamicDowncast<SVGElement>(parentNode()); svgParent && svgParent->isTextContent())
```
This is hard to read. Can you please move the casting out of the if-statement.

cdumez replied:
Then the scope of the variable is way larger than needed, plus it's a raw pointer. I'll make the change since you're maintaining SVG but I think my code was better.

shallawa commented on Source/WebCore/svg/SVGLocatable.cpp:
```diff
@@ -25,29 +25,37 @@
 namespace WebCore {
 
-static bool isViewportElement(Node* node)
+static SVGElement* toViewportElement(Element& element)
```
I think this is a little bit confusing. Someone would expect `toViewportElement` has to convert the `Element` to `ViewportElement` which does not exist. Also SVGElement has a method called `SVGElement::viewportElement()`.

cdumez replied:
I don't think I made it much worse given that the original name was `isViewportElement()`. I'd really like to address in this PR so I'm open to suggestions on a better name.

cdumez replied:
I updated my PR, I found a way to keep isViewportElement() mostly as-is while still optimizing what I wanted to.

shallawa commented on Source/WebCore/svg/SVGLocatable.cpp:
```diff
@@ -25,29 +25,37 @@
+    return nullptr;
 }
 
 SVGElement* SVGLocatable::nearestViewportElement(const SVGElement* element)
```
`SVGElement::viewportElement()` is very similar to this function except it does not check for foreign object. I would suggest leaving this method as is for now and address the clarity and the code redundancy in a separate PR.

cdumez committed:
Reduce use of downcast<>() in SVG code
https://bugs.webkit.org/show_bug.cgi?id=266678

Reviewed by Simon Fraser.

Reduce use of downcast<>() in SVG code. Use dynamicDowncast<>() instead.

* Source/WebCore/svg/DocumentSVG.cpp:
(WebCore::DocumentSVG::rootElement):
* Source/WebCore/svg/SVGAElement.cpp:
(WebCore::SVGAElement::createElementRenderer):
(WebCore::SVGAElement::defaultEventHandler):
* Source/WebCore/svg/SVGAltGlyphDefElement.cpp:
(WebCore::SVGAltGlyphDefElement::hasValidGlyphElements const):
* Source/WebCore/svg/SVGAltGlyphElement.cpp:
(WebCore::SVGAltGlyphElement::hasValidGlyphElements const):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::shouldApplyPathClipping const):
* Source/WebCore/svg/SVGComponentTransferFunctionElement.cpp:
(WebCore::SVGComponentTransferFunctionElement::svgAttributeChanged):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::correspondingUseElement const):
(WebCore::SVGElement::childShouldCreateRenderer const):
(WebCore::SVGElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::buildPendingResource):
* Source/WebCore/svg/SVGFELightElement.cpp:
(WebCore::SVGFELightElement::svgAttributeChanged):
* Source/WebCore/svg/SVGFilterElement.cpp:
(WebCore::SVGFilterElement::childShouldCreateRenderer const):
* Source/WebCore/svg/SVGFilterPrimitiveStandardAttributes.h:
(isType):
* Source/WebCore/svg/SVGFontFaceElement.cpp:
(WebCore::SVGFontFaceElement::rebuildFontFace):
* Source/WebCore/svg/SVGFontFaceFormatElement.cpp:
(WebCore::SVGFontFaceFormatElement::childrenChanged):
* Source/WebCore/svg/SVGFontFaceSrcElement.cpp:
(WebCore::SVGFontFaceSrcElement::childrenChanged):
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:
(WebCore::SVGFontFaceUriElement::childrenChanged):
* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
(WebCore::SVGGeometryElement::isPointInFill):
(WebCore::SVGGeometryElement::isPointInStroke):
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::toViewportElement):
(WebCore::SVGLocatable::nearestViewportElement):
(WebCore::SVGLocatable::farthestViewportElement):
(WebCore::isViewportElement): Deleted.
* Source/WebCore/svg/SVGTextPathElement.cpp:
(WebCore::SVGTextPathElement::buildPendingResource):

Canonical link: https://commits.webkit.org/272374@main
